File: roadhelp_app/lib/models/issue.dart
import 'dart:convert';

import '/config/enums.dart';
import '/models/rating_issue.dart';
import '/models/user.dart';
import 'base_model.dart';

class Issue extends BaseModel {
  //
  int? userMemberId;
  int? userPartnerId;

  //
  String? address;
  String? phone;
  double? longitude;
  double? latitude;
  IssueCategory? category;
  String? description;
  IssueStatus? status;

  //Relationship
  User? userMember;
  User? userPartner;
  RatingIssue? ratingIssue;

  Issue({
    this.userMemberId,
    this.userPartnerId,
    this.address,
    this.phone,
    this.longitude,
    this.latitude,
    this.category,
    this.description,
    this.status,
    //Relationship
    this.userMember,
    this.userPartner,
    this.ratingIssue,
    //
    int? id,
    DateTime? createdAt,
    String? createdBy,
    DateTime? updatedAt,
    String? updatedBy,
    int? version,
    bool? deleted,
  }) : super(
          id: id,
          createdAt: createdAt,
          createdBy: createdBy,
          updatedAt: updatedAt,
          updatedBy: updatedBy,
          version: version,
          deleted: deleted,
        );

  //
  factory Issue.fromJson(Map<String, dynamic> json) {
    return Issue(
      userMemberId: json['userMemberId'],
      userPartnerId: json['userPartnerId'],
      //
      longitude: json['longitude'],
      latitude: json['latitude'],
      phone: json['phone'],
      address: json['address'],
      category: IssueCategory.values.firstWhere(
          (e) => e.toString() == "IssueCategory." + json['category']),
      description: json['description'],
      status: IssueStatus.values
          .firstWhere((e) => e.toString() == "IssueStatus." + json['status']),
      //Relationship
      userMember:
          json['userMember'] != null ? User.fromJson(json['userMember']) : null,
      userPartner: json['userPartner'] != null
          ? User.fromJson(json['userPartner'])
          : null,
      ratingIssue: json['ratingIssue'] != null
          ? RatingIssue.fromJson(json['ratingIssue'])
          : null,
      //
      id: json['id'],
      createdAt: json['createdAt'] != null ? DateTime.fromMillisecondsSinceEpoch(json['createdAt'], isUtc: true) : null,
      createdBy: json['createdBy'],
      updatedAt: json['updatedAt'] != null ? DateTime.fromMillisecondsSinceEpoch(json['updatedAt'], isUtc: true) : null,
      updatedBy: json['updatedBy'],
      version: json['version'],
      deleted: json['deleted'],
    );
  }

  //
  static List<Issue> fromJsonToList(dynamic json) {
    return json.map<Issue>((element) => Issue.fromJson(element)).toList();
  }

  //
  String toJson() {
    return json.encode({
      'userMemberId': userMemberId,
      'userPartnerId': userPartnerId,
      //
      'address': address,
      'phone': phone,
      'longitude': longitude,
      'latitude': latitude,
      'category': category != null ? category!.index : null,
      'description': description,
      'status': status != null ? status!.index : null,
      //Relationship
      'userMember':
          userMember != null ? json.decode(userMember!.toJson()) : null,
      'userPartner':
          userPartner != null ? json.decode(userPartner!.toJson()) : null,
      'ratingIssue':
          ratingIssue != null ? json.decode(ratingIssue!.toJson()) : null,
      //
      'id': id,
      //'createdAt': createdAt,
      'createdBy': createdBy,
      //'updatedAt': updatedAt, //DateTime.now().millisecondsSinceEpoch | updatedAt.millisecondsSinceEpoch
      'updatedBy': updatedBy,
      'version': version,
      'deleted': deleted,
    });
  }
}

File: roadhelp_web/src/main/java/com/codedy/roadhelp/model/Issue.java
package com.codedy.roadhelp.model;

import com.codedy.roadhelp.model.enums.IssueCategory;
import com.codedy.roadhelp.model.enums.IssueStatus;

import javax.persistence.*;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.util.LinkedHashMap;

@Entity
@Table(name = "issue")
public class Issue extends BaseModel implements Serializable {

    //region - Define Fields -
    private double longitude;

    private double latitude;

    @Size(min = 10)
    private String phone;

    @Size(max = 128)
    private String address;

    @Size(max = 500)
    private String description;

    private IssueCategory category;

    private IssueStatus status;
    //endregion


    //region - Relationship -
    @ManyToOne
    @JoinColumn(name = "user_member_id")
    private User userMember;

    @ManyToOne
    @JoinColumn(name = "user_partner_id")
    private User userPartner;

    @OneToOne(mappedBy = "issue")
    private RatingIssue ratingIssue;
    //endregion


    //region - Getter & Setter -
    public IssueCategory getCategory() {
        return category;
    }

    public void setCategory(IssueCategory category) {
        this.category = category;
    }

    public User getUserMember() {
        return userMember;
    }

    public void setUserMember(User userMember) {
        this.userMember = userMember;
    }

    public User getUserPartner() {
        return userPartner;
    }

    public void setUserPartner(User userPartner) {
        this.userPartner = userPartner;
    }

    public RatingIssue getRatingIssues() {
        return ratingIssue;
    }

    public void setRatingIssues(RatingIssue ratingIssue) {
        this.ratingIssue = ratingIssue;
    }

    public RatingIssue getRatingPartner() {
        return ratingIssue;
    }

    public void setRatingPartner(RatingIssue ratingIssue) {
        this.ratingIssue = ratingIssue;
    }

    public IssueStatus getStatus() {
        return status;
    }

    public void setStatus(IssueStatus status) {
        this.status = status;
    }

    public double getLongitude() {
        return longitude;
    }

    public void setLongitude(double longitude) {
        this.longitude = longitude;
    }

    public double getLatitude() {
        return latitude;
    }

    public void setLatitude(double latitude) {
        this.latitude = latitude;
    }

    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

//endregion


    //region - Relationship Helper -

    /**
     * Hàm này trả về cấu trúc nguyên thủy của entity này.<br/><br/>
     * <p>
     * Viết bởi: Hiếu iceTea<br/>
     * Ngày: 23-10-2021<br/>
     * Thời gian: 22:22<br/>
     *
     * @return
     */
    protected LinkedHashMap<String, Object> toHashMap() {
        LinkedHashMap<String, Object> hashMap = super.toHashMap();

        hashMap.put("userMemberId", userMember != null ? userMember.getId() : null);
        hashMap.put("userPartnerId", userPartner != null ? userPartner.getId() : null);

        hashMap.put("longitude", longitude);
        hashMap.put("latitude", latitude);
        hashMap.put("phone", phone);
        hashMap.put("address", address);
        hashMap.put("category", category);
        hashMap.put("description", description);
        hashMap.put("status", status);

        return hashMap;
    }

    /**
     * Hàm này trả về định dạng hiển thị dữ liệu API cho entity này.<br/><br/>
     * <p>
     * Viết bởi: Hiếu iceTea<br/>
     * Ngày: 24-10-2021<br/>
     * Thời gian: 10:15<br/>
     *
     * @return
     */
    public LinkedHashMap<String, Object> toApiResponse() {
        LinkedHashMap<String, Object> hashMap = this.toHashMap();

        hashMap.remove("userMemberId");
        hashMap.remove("userPartnerId");

        hashMap.put("userMember", getUserMemberHashMap());
        hashMap.put("userPartner", getUserPartnerHashMap());
        hashMap.put("ratingIssue", getRatingIssueHashMap());

        return hashMap;
    }

    //@JsonProperty("userMember")
    private LinkedHashMap<String, Object> getUserMemberHashMap() {
        return userMember != null ? userMember.toHashMap() : null;
    }

    //@JsonProperty("userPartner")
    private LinkedHashMap<String, Object> getUserPartnerHashMap() {
        return userPartner != null ? userPartner.toHashMap() : null;
    }

    //@JsonProperty("ratingIssue")
    private LinkedHashMap<String, Object> getRatingIssueHashMap() {
        return ratingIssue != null ? ratingIssue.toHashMap() : null;
    }
    //endregion

}
